### PR TITLE
Swizzle

### PR DESCRIPTION
**Pull request by calebzulawski**
This replaces the "shuffle" API with a more generic implementation.  With these changes, you can now create any length vector from any other length vector (a "swizzle").

The API isn't _great_ for creating arbitrary, one-off shuffles (e.g. `_mm_shuffle_*`), but it's great for algorithmically generating shuffles.  We could consider something like a macro for one-off shuffles, but not sure if that's a great idea, I left that for later.

One thing I don't love in this PR is the `Which` enum, at minimum I think it could use a better name.

### CODE REVIEW

**Comment by thomcc** [comment]
So you have to define a dummy struct and implement a trait for each swizzle? That doesn't seem great...

**Comment by calebzulawski** [comment]
Well, it kind of is a dummy struct I suppose, but traits are the only way right now to have a dependency on a const that you can parameterize.

**Comment by calebzulawski** [comment]
To clarify, there is no way to do `<T, const V: T>`

**Comment by Lokathor** [comment]
This is... a remarkably poor direction to take.

**Comment by calebzulawski** [comment]
Well, I am open to any other ideas, but I don't think there is _any_ other way to do it, other than with a trait const.

It's worth noting that 1) this allowed implementation of `rotate_{left,right}` which was previously not possible, and 2) you can still do something like this for convenience:

```
pub struct Shuffle4<const I: [usize; 4]>;

impl<const I: [usize; 4]> Swizzle<4, 4> for Shuffle4<I> {
    const INDEX: [usize; 4] = I;
}
```

**Comment by workingjubilee** [comment]
> Well, I am open to any other ideas, but I don't think there is any other way to do it, without a trait const.

A "trait const"? Do you mean a trait-level const generic parameter?
As I have mentioned before, I do not think that is the worst sin.
The code using a trait could ideally, for a user, look something like

```rust
use core::simd::{f32x4, f32x8, Swizzle};

fn get_prime_indices(floats: f32x8) -> f32x4 {
  Swizzle::swizzle<{ [2, 3, 5, 7] }>(floats)
}
```

That's fine? I haven't reviewed this code as closely as I would like but it does seem... slightly heavier weight than that.

**Comment by calebzulawski** [comment]
I meant associated const.  Unfortunately I don't think there's any way to do what you're suggesting, because there is a const panic to catch invalid indices, as well as an index conversion in the case of the 2-argument swizzle.

Even if you forgo avoiding monomorphisation errors, you still can't pass the array in via a generic parameter, because the index type can't depend on outer parameters.

**Comment by Lokathor** [comment]
isn't the index type always usize? or, is it sometimes u32 instead or something?

can't we make it always u32? i mean it's a const so the actual thing is gonna usually be literals anyway.

**Comment by programmerjake** [comment]
> isn't the index type always usize? or, is it sometimes u32 instead or something?
> 
> can't we make it always u32? i mean it's a const so the actual thing is gonna usually be literals anyway.

I think what @calebzulawski meant was that the array length of the index type was the problem, not `u32` vs. `usize`.

**Comment by calebzulawski** [comment]
The usize isn't the issue--I meant the entire type, including the length.  You cannot have generic parameters depend on other generic parameters, which prevents doing anything like `Swizzle::<N>::swizzle<[0; N]>()`.  There is no way to create a function that has the array as a const generic parameter, since the length of the array is generic as well.

The reason this implementation works is because the array type is _effectively concrete_ once the trait is monomorphised, and we can then create regular consts (such as `Swizzle::INDEX`) with that concrete type.

File: crates/core_simd/src/swizzle.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+/// Create a vector from the elements of another vector.
+pub trait Swizzle<const INPUT_LANES: usize, const OUTPUT_LANES: usize> {
+    /// Map from the lanes of the input vector to the output vector.
+    const INDEX: [usize; OUTPUT_LANES];
+
+    #[doc(hidden)]
+    /// The `simd_shuffle` intrinsic expects `u32`, so do error checking and conversion here.
+    const INDEX_IMPL: [u32; OUTPUT_LANES] = check::<INPUT_LANES, OUTPUT_LANES>(Self::INDEX);
```
Comment (programmerjake) [inline crates/core_simd/src/swizzle.rs:20]: this should be somehow protected, since a user can currently cause UB without `unsafe` by overriding `INDEX_IMPL`.

File: crates/core_simd/src/swizzle.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+/// Create a vector from the elements of two other vectors.
+pub trait Swizzle2<const INPUT_LANES: usize, const OUTPUT_LANES: usize> {
+    /// Map from the lanes of the input vectors to the output vector
+    const INDEX: [Which; OUTPUT_LANES];
+
+    #[doc(hidden)]
+    /// The `simd_shuffle` intrinsic expects `u32`, so do error checking and conversion here.
+    const INDEX_IMPL: [u32; OUTPUT_LANES] = check2::<INPUT_LANES, OUTPUT_LANES>(Self::INDEX);
```
Comment (programmerjake) [inline crates/core_simd/src/swizzle.rs:42]: this also needs protection

File: crates/core_simd/src/swizzle.rs (hunk 1 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:20]: Actually, it wouldn't cause UB, the compiler would bail with a monomorphisation error if we'rw ok with that.

Comment (programmerjake) [reply crates/core_simd/src/swizzle.rs:20]: we could do something like:
```rust
mod private {
pub trait SwizzleConsts<const A: usize, const B: usize>: Swizzle<A, B> {
    const INDEX_IMPL: [...; ...];
}
}

pub trait Swizzle<const A: usize, const B: usize> {
    const INDEX: [...; ...];
    fn swizzle(self) -> ...
    where
        Self: private::SwizzleConsts<A, B>,
    {
        ...
    }
}

impl<T, const A: usize, const B: usize> private::SwizzleConsts<A, B> for T
where
    T: Swizzle<A, B>,
{
    ...
}
```

Comment (thomcc) [reply crates/core_simd/src/swizzle.rs:20]: Inside the stdlib this stuff often can be protected using `#[unstable]`, even if its a bit of a hack.

**Comment by calebzulawski** [comment]
Thinking a little more about this--I still think this trait is necessary.  The concern about the ergonomics of a dummy struct is valid, but I think the ergonomics it provides are absolutely necessary:
* Error check lengths, panicking on overflowing the input vector, much more desirable than the modulo arithmetic LLVM provides.  It's really hard to debug your index calculation when your off-by-one error just turns into a rotation instead of a hard error.
* Using an enum to select between your input vectors in the case of a 2-vector swizzle. Manually calculating the offset into the second vector is a terrible API.

This simply isn't possible with fn generics.  Even if we extended const generics, you still won't be able to do:
```rust
fn shuffle<T, const LANES: usize, const INDEX: [usize; LANES]>(self) -> Self {
    const _ASSERT: () = our_assert_fn(INDEX);
}
```
for the same reason that you can't do:
```rust
fn foo<T>() {
    const FOO: Option<T> = None;
}
```
It's not a limitation of const generics, it's just generics and monomorphization in general.

I propose that we include this trait for now, which isn't particularly convenient to use (but really not that bad, either).  It's usable _today_, even on stable.  In the future, _if_ const generics are extended this way, we can then provide the following if we deem it necessary:
```rust
struct Shuffle<const LANES: usize, const INDEX: [usize; LANES]>;
struct Shuffle2<const LANES: usize, const INDEX: [Which; LANES]>;
```
This would allow you to write something like `Shuffle::<4, [4, 3, 2, 1]>::swizzle(x)`, which would cover the common cases.

**Comment by thomcc** [comment]
> It's not a limitation of const generics, it's just generics and monomorphization in general.

I think there's a workaround for this that gets used in e.g. stdarch, see: https://github.com/rust-lang/stdarch/blob/master/crates/core_arch/src/macros.rs, which is used to statically assert things about const generic parameters.

File: crates/core_simd/src/swizzle.rs (hunk 1 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs]: I was able to hide the impls with a helper trait like this

**Comment by calebzulawski** [comment]
Hmm, that does look like it might be a good trick to get past the monomorphization problems, I'd amend my plan to say if we get the hypothetical const generics extension developed and stabilized before we need shuffle stabilized, we try that out.

**Comment by programmerjake** [comment]
> ```rust
> struct Shuffle<const LANES: usize, const INDEX: [usize; LANES]>;
> struct Shuffle2<const LANES: usize, const INDEX: [Which; LANES]>;
> ```

Well, now that I think about it, wouldn't it be better to do something like:
```rust
pub struct Shuffle<const V: &[usize]>;
pub struct Shuffle2<const V: &[Which]>;
```
Which allows doing something like `Shuffle::<&[4, 3, 2, 1]>::swizzle(x)`

**Comment by calebzulawski** [comment]
Putting aside that you can't use slices as const generic parameters right now, one of the downsides of that is it makes the type deduction on the return type pretty nasty.  Definitely worth considering as const generics progress, though.

**Comment by calebzulawski** [comment]
I just added a `simd_shuffle` macro, which I believe handles the API everyone seems to be looking for in the common case.  It's a macro, not a function, but again it works on current rust.  I was thinking we could address this in a later PR but it seems to be a big enough priority that I just took care of it.  It has all of the ergonomics benefits this PR already had, but minimizes the boilerplate in the common case of a regular shuffle.  I still think we need to provide the full shuffle interface for more complex things (see the implementation of `rotate_{left,right}` as an example...), but this is much less verbose for common usage.

**Comment by calebzulawski** [comment]
I think the 2-input-shuffle in the matrix inversion example shows how much more readable it is specifying which vector you are choosing from, especially with all the noise gone when using the macro.

File: crates/core_simd/src/swizzle.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+#[macro_export]
+macro_rules! simd_shuffle {
+    {
+        $vector:expr, $index:expr $(,)?
+    } => {
+        {
+            // FIXME this won't work when we are in `core`!
+            use $crate::Swizzle;
+            struct Shuffle;
+            impl Swizzle<{$index.len()}, {$index.len()}> for Shuffle {
```
Comment (programmerjake) [inline crates/core_simd/src/swizzle.rs:42]: this should be generic over input vector length

File: crates/core_simd/src/swizzle.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+        }
+    };
+    {
+        $first:expr, $second:expr, $index:expr $(,)?
+    } => {
+        {
+            // FIXME this won't work when we are in `core`!
+            use $crate::{Which, Swizzle2};
+            struct Shuffle;
+            impl Swizzle2<{$index.len()}, {$index.len()}> for Shuffle {
```
Comment (programmerjake) [inline crates/core_simd/src/swizzle.rs:55]: this should also be generic over input vector length

File: crates/core_simd/src/swizzle.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+trait SwizzleImpl<const INPUT_LANES: usize, const OUTPUT_LANES: usize> {
+    const INDEX_IMPL: [u32; OUTPUT_LANES];
+}
+
+impl<T, const INPUT_LANES: usize, const OUTPUT_LANES: usize> SwizzleImpl<INPUT_LANES, OUTPUT_LANES>
+    for T
+where
+    T: Swizzle<INPUT_LANES, OUTPUT_LANES> + ?Sized,
+{
+    const INDEX_IMPL: [u32; OUTPUT_LANES] = {
```
Comment (programmerjake) [inline crates/core_simd/src/swizzle.rs:122]: looks good!

**Comment by workingjubilee** [comment]
cc @miguelraz, I am genuinely curious what he thinks of this sample since he ported the Matrix Inversion code.
both the overall diff:
https://github.com/rust-lang/portable-simd/pull/161/files/109b908c071dd082a1c23706f5951a6c2330e8b1
and especially the diff here, between the pre/post macro version:
https://github.com/rust-lang/portable-simd/pull/161/commits/109b908c071dd082a1c23706f5951a6c2330e8b1

**Comment by Lokathor** [comment]
> const SHUFFLE01: [Which; 4] = [First(0), First(1), Second(0), Second(1)];

While I maybe understand, based on crusty knowledge about how the sse shuffles work, why this looks like this, it's very opaque.

**Comment by calebzulawski** [comment]
> > const SHUFFLE01: [Which; 4] = [First(0), First(1), Second(0), Second(1)];
> 
> While I maybe understand, based on crusty knowledge about how the sse shuffles work, why this looks like this, it's very opaque.

Maybe I'm forgetting something, but SSE/AVX doesn't have anything like this interface.  I think you would have to do two shuffles and a blend.

I'm certainly opening to rewording etc, but `[0, 1, 5, 6]` seems much worse to me.  First of all, it's limited to use only with length-4 inputs, and secondly you need to _remember_ that the input is length 4 to even read it, and finally you need to actually do the modulo correctly in your head when reading it.  I definitely prefer if both vectors are 0-indexed and the length isn't encoded into the second vector index.

**Comment by calebzulawski** [comment]
Proof to my point above, the equivalent would be `[0, 1, 4, 5]` and I screwed up the math 🤦

**Comment by Lokathor** [comment]
i presumed it was based on https://doc.rust-lang.org/nightly/core/arch/x86/fn._mm_shuffle_ps.html, which shuffles two vectors into one output vector.

And the fact that you also kinda messed up the math i think shows that it's maybe very weird.

it might be fixable with sufficient macro power though

**Comment by calebzulawski** [comment]
I suppose I'm confused what the problem is. I messed up the math because I screwed up the modulo in my head, which is why I'm suggesting we don't use the modulo representation.  It's very simple, `First(i)` takes `first[i]` and `Second(i)` takes `second[i]`.  Rather than the LLVM representation of `i` takes `if i < first.len() { first[I] } else { second[i - first.len()] }`

**Comment by calebzulawski** [comment]
This is very different from the SSE instruction that takes half from the first vector and half from the second vector.  With this API you can do `[First(0), Second(0), First(0), Second(0)]` which can't be done with that SSE instruction.

**Comment by Lokathor** [comment]
Ah, see that's the confusion I guess. It looks similar enough to sse shuffle that i tried to decode it with my sse brain.

**Comment by calebzulawski** [comment]
I think it will look similar to all shuffles, which have their own quirks (neon's vtbl/vtbx are a little weird too).  I think this is the API with the least magic, you get exactly the lanes you ask for, and anything else will error.

File: crates/core_simd/src/swizzle.rs (hunk 3 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:42]: That would be desirable, but I don't see a way of extracting the const length from a non-const vector expression.  It's probably possible, so I'll have to think about it a bit more.

File: crates/core_simd/src/swizzle.rs [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+            }
+            Shuffle::swizzle($vector)
+        }
+    };
+    {
+        $first:expr, $second:expr, $index:expr $(,)?
+    } => {
+        {
+            // FIXME this won't work when we are in `core`!
+            use $crate::{Which, Swizzle2};
```
Comment (workingjubilee) [inline crates/core_simd/src/swizzle.rs:53]:
```suggestion
            use $crate::simd::Swizzle;
            struct Shuffle;
            impl Swizzle<{$index.len()}, {$index.len()}> for Shuffle {
                const INDEX: [usize; {$index.len()}] = $index;
            }
            Shuffle::swizzle($vector)
        }
    };
    {
        $first:expr, $second:expr, $index:expr $(,)?
    } => {
        {
            use $crate::simd::{Which, Swizzle2};
```

Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs]: I just remembered my original issue.  `core_simd::simd` isn't public, so a user-accessible macro can't access it.  I'm just going to make it public.

File: crates/core_simd/src/swizzle.rs (hunk 3 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs]: I realized I was overthinking it--the implementation was easy.  I changed the macro to `simd_swizzle` to reflect that, too.

**Comment by programmerjake** [comment]
maybe it'd also be a good idea to rename the rustc `simd_shuffle` intrinsics to `simd_swizzle` for consistency's sake

**Review by programmerjake — APPROVED** [review]
lgtm

**Comment by calebzulawski** [comment]
To note something that came up in rust-lang/rust#89167, this change also removes the requirement to use the `adt_const_params` feature, which is incomplete.

**Comment by workingjubilee** [comment]
Finally properly reviewed this. There's a lot of duplication in our API and its appearance in the documentation but for some reason the `simd_swizzle!` macro looks like it's in the namespace twice. Do you have any idea why that is, and if it can be removed?

File: crates/core_simd/src/swizzle.rs [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+/// An index into one of two vectors.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Which {
+    /// Indexes the first vector.
+    First(usize),
+    /// Indexes the second vector.
+    Second(usize),
+}
```
Comment (workingjubilee) [inline crates/core_simd/src/swizzle.rs:80]: It is traditional to name this type "Either", and its variants are named "Left" and "Right".
I am not a particularly devout adherent of tradition, however, so I am fine with whatever, at the moment.

File: crates/core_simd/src/swizzle.rs [hunk 8]
```diff
@@ -0,0 +1,374 @@
+use crate::simd::intrinsics;
+use crate::{LaneCount, Simd, SimdElement, SupportedLaneCount};
+
+/// Rearrange vector elements.
+///
+/// A new vector is constructed by specifying the the lanes of the source vector or vectors to use.
```
Comment (workingjubilee) [inline crates/core_simd/src/swizzle.rs:6]: This should prefer active voice and possibly less repetition, e.g.
```suggestion
/// Constructs a new vector by selecting values from the lanes of the source vector or vectors to use.
```

File: crates/core_simd/src/swizzle.rs [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+        T: SimdElement,
+        LaneCount<INPUT_LANES>: SupportedLaneCount,
+        LaneCount<OUTPUT_LANES>: SupportedLaneCount,
+    {
+        unsafe { intrinsics::simd_shuffle(first, second, Self::INDEX_IMPL) }
+    }
+}
+
+/// The `simd_shuffle` intrinsic expects `u32`, so do error checking and conversion here.
+trait SwizzleImpl<const INPUT_LANES: usize, const OUTPUT_LANES: usize> {
```
Comment (workingjubilee) [inline crates/core_simd/src/swizzle.rs:123]: Okay, I am actually kinda confused why we have both these and the other two Swizzle traits? "SwizzleImpl" isn't telling me much, as both traits are implemented. This is trying to say something about being used to define the indexing behavior.

File: crates/core_simd/src/swizzle.rs [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+
+    /// Deinterleave two vectors.
+    ///
+    /// The first result takes every other lane of `self` and then `other`, starting with
+    /// the first lane.
+    ///
+    /// The second result takes every other lane of `self` and then `other`, starting with
+    /// the second lane.
+    ///
+    /// This particular permutation is efficient on many architectures.
```
Comment (workingjubilee) [inline crates/core_simd/src/swizzle.rs:327]: This is very vague, now that I look at it.
Is this about the architecture itself (as in x86 or aarch64) or the microarchitecture (as in Haswell vs. Nehalem)?

File: crates/core_simd/src/swizzle.rs [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+            }
+            SwizzleImpl::swizzle($vector)
+        }
+    };
+    {
+        $first:expr, $second:expr, $index:expr $(,)?
+    } => {
+        {
+            use $crate::simd::{Which, Swizzle2};
+            struct SwizzleImpl;
```
Comment (workingjubilee) [inline crates/core_simd/src/swizzle.rs:64]: Here we define a struct that shares a name with a trait later...

**Review by workingjubilee — COMMENTED** [review]
I have reviewed the code itself for correctness and I am relatively confident of it but it is hard to follow what is going on here in a few places. Additional commentary for *ourselves* would be useful, even if it might just seem like nattering on to users.

File: crates/core_simd/src/swizzle.rs (hunk 7 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:78]: Agreed, now that I look at it, it seems like this really should just be an `Either`.  Which isn't really the place of `core::simd` to add, but it's by far a superior API in this case.

File: crates/core_simd/src/swizzle.rs (hunk 9 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:121]: I'll try to improve the comments.  This trait is automatically implemented, it exists purely to create `INDEX_IMPL`, which is `u32` instead of `usize`, and may panic upon evaluation.

Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:121]: (it's a separate trait to keep `INDEX_IMPL` completely hidden from the user)

File: crates/core_simd/src/swizzle.rs (hunk 10 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:325]: I think it's architecture--x86 has both generic shuffle and unpack, and aarch64 has unzip.  I think it's intentionally vague, as it's trying to just say "this shuffle is probably okay to use frequently".

File: crates/core_simd/src/swizzle.rs (hunk 7 above)
Comment (workingjubilee) [reply crates/core_simd/src/swizzle.rs:78]: Normally in Rust we use Result for what is called Either, because in languages that have Either, they implicitly use it as a Result, but sometimes it really is Either. So I think it is harmless if we have `core::simd::Either`: it's specifically for this library, after all.

**Comment by workingjubilee** [comment]
> Finally properly reviewed this. There's a lot of duplication in our API and its appearance in the documentation but for some reason the `simd_swizzle!` macro looks like it's in the namespace twice. Do you have any idea why that is, and if it can be removed?

Might be a `macro_use` combined with a `use module::*;` or something like that.
I don't think this is essential to fix, heh, so it's not at all a blocker, it's just something I have started paying attention to.

File: crates/core_simd/src/swizzle.rs (hunk 10 above)
Comment (workingjubilee) [reply crates/core_simd/src/swizzle.rs:325]: Okay, we should probably say "architectures have specialized instructions for this" then.

**Comment by workingjubilee** [comment]
Also re-reviewing again and I just want to affirm I will be happy with this when the internal documentation/commenting concerns are addressed so I can check things against what I thought things were for. ^^;

File: crates/core_simd/src/swizzle.rs (hunk 11 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs]: Changed the name to make it a little less confusing

File: crates/core_simd/src/swizzle.rs (hunk 7 above)
Comment (calebzulawski) [reply crates/core_simd/src/swizzle.rs:78]: I think maybe it's not quite correct to call it `Either`, since that would imply that it should be `Either<usize>`, I think.

**Comment by calebzulawski** [comment]
Worth noting that this is unique to `core_simd`, and won't occur in `core` itself.

File: crates/core_simd/src/swizzle.rs (hunk 7 above)
Comment (workingjubilee) [reply crates/core_simd/src/swizzle.rs:78]: That would be `Maybe<usize>`, no?
I am not of a strong opinion however, here.
